### PR TITLE
Introduce Integration Signature Type Array

### PR DESCRIPTION
**Pull request by colin-higgins**
Changes proposed in this pull request:
- [x] Enforce type signature array specification with test
- [x] Decorate every wrapper method with empty type signature array
- [x] Serialize type signature array into integrations.json
- [x] Deserialize type signature array into c++
- [x] Write tests verifying deserialization behavior
- [x] Introduce short-circuits based on signature type parsing

@DataDog/apm-dotnet

### CODE REVIEW

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AdoNetIntegration.cs [hunk 1]
```diff
@@ -162,7 +166,7 @@ private static string GetDbType(string commandTypeName)
                 case "InterceptableDbCommand":
                 case "ProfiledDbCommand":
                     // don't create spans for these
-                    return null;
+                    return TypeNames.Ignore;
```
Comment (lucaspimentel) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/AdoNetIntegration.cs]: This is unrelated and should still return `null`.

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/Types/TypeNames.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+namespace Datadog.Trace.ClrProfiler
+{
+    internal class TypeNames
```
Comment (lucaspimentel) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/Types/TypeNames.cs]: nit: this class can be static

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/Types/TypeNames.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        public const string Ignore = "_";
+
+        public const string Void = "System.Void";
+        public const string Object = "System.Object";
+        public const string Bool = "System.Boolean";
+
+        public const string Int8 = "System.Int8";
+        public const string Int16 = "System.Int8";
```
Comment (lucaspimentel) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/Types/TypeNames.cs]:
```suggestion
        public const string Int16 = "System.Int16";
```

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/Types/TypeNames.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        public const string Int32 = "System.Int32";
+        public const string Int64 = "System.Int64";
+
+        public const string UInt8 = "System.UInt8";
+        public const string UInt16 = "System.UInt8";
```
Comment (lucaspimentel) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/Types/TypeNames.cs]:
```suggestion
        public const string UInt16 = "System.UInt16";
```

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/Types/TypeNames.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        public const string UInt32 = "System.UInt32";
+        public const string UInt64 = "System.UInt64";
+
+        public const string Byte = Int8;
```
Comment (lucaspimentel) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/Types/TypeNames.cs]:
```suggestion
        public const string Byte = UInt8;
```

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp [hunk 6]
```diff
@@ -442,9 +442,37 @@ HRESULT STDMETHODCALLTYPE CorProfiler::JITCompilationStarted(
...
+      }
+
+      if (!is_match) {
+        continue;
+      }
+
       const auto original_argument = pInstr->m_Arg32;
 
-      // insert the opcode and signature token as
+      // insert the op-code and signature token as
```
Comment (lucaspimentel) [inline src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: https://en.wikipedia.org/wiki/Opcode 🤔

File: src/Datadog.Trace.ClrProfiler.Native/integration_loader.cpp [hunk 7]
```diff
@@ -136,24 +136,39 @@ MethodReference MethodReferenceFromJson(const json::value_type& src) {
...
+    }
+    if (src.find("maximum_minor") != eoj) {
+      max_minor = src["maximum_minor"].get<USHORT>();
+    }
+    if (src.find("maximum_patch") != eoj) {
+      max_patch = src["maximum_patch"].get<USHORT>();
+    }
+
+    if (src.find("signature_types") != eoj) {
+      // nlohmann is unable to handle null values in this array
```
Comment (lucaspimentel) [inline src/Datadog.Trace.ClrProfiler.Native/integration_loader.cpp]: 😝 
```suggestion
      // C++ is unable to handle null values in this array
```

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/Types/TypeNames.cs (hunk 5 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/Types/TypeNames.cs]: Shouldn't this be "System.Byte"?

File: integrations.json [hunk 8]
```diff
@@ -468,6 +568,13 @@
           "assembly": "ServiceStack.Redis",
           "type": "ServiceStack.Redis.RedisNativeClient",
           "method": "SendReceive",
+          "signature_types": [
+            "T",
+            "_",
+            "System.Int8[][]",
```
Comment (zacharycmontoya) [inline integrations.json]:
```suggestion
            "System.Byte[][]",
```

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/Types/TypeNames.cs [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+{
+    internal class TypeNames
+    {
+        public const string Ignore = "_";
+
+        public const string Void = "System.Void";
+        public const string Object = "System.Object";
+        public const string Bool = "System.Boolean";
+
+        public const string Int8 = "System.Int8";
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/Types/TypeNames.cs]:
```suggestion
        public const string SByte = "System.SByte";
```

File: integrations.json (hunk 8 above)
Comment (zacharycmontoya) [reply integrations.json]: Nvm this will be regenerated. I'll put the suggestion in the right place in the TargetSignatureTypes...

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/ServiceStackRedisIntegration.cs [hunk 10]
```diff
@@ -28,6 +28,7 @@ public static class ServiceStackRedisIntegration
             CallerAssembly = "ServiceStack.Redis",
             TargetAssembly = "ServiceStack.Redis",
             TargetType = "ServiceStack.Redis.RedisNativeClient",
+            TargetSignatureTypes = new[] { "T", TypeNames.Ignore, "System.Int8[][]", TypeNames.Ignore, TypeNames.Bool },
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/ServiceStackRedisIntegration.cs]:
```suggestion
            TargetSignatureTypes = new[] { "T", TypeNames.Ignore, "System.Byte[][]", TypeNames.Ignore, TypeNames.Bool },
```

Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/ServiceStackRedisIntegration.cs]: Wait, shouldn't the double array be the first argument after the return type "T"?

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/Types/TypeNames.cs (hunk 5 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/Types/TypeNames.cs]: Also, you should be able to collapse UInt8 into Byte because `System.Byte` is the struct for an unsigned 8-bit integer.

Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/Types/TypeNames.cs]: Probably. Maybe we should also drop `System.Int8` and `System.UInt8` since those don't exist as .NET types.

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/Types/ClrNames.cs [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+        public const string Int16 = "System.Int16";
+        public const string Int32 = "System.Int32";
+        public const string Int64 = "System.Int64";
+
+        public const string UInt8 = "System.UInt8";
+        public const string UInt16 = "System.UInt16";
+        public const string UInt32 = "System.UInt32";
+        public const string UInt64 = "System.UInt64";
+
+        public const string Byte = UInt8;
```
Comment (lucaspimentel) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/Types/ClrNames.cs]: I'm thinking since we are using the BCL names here (with the `System` namespace), we should change `UInt8` to `Byte` and `Int8` to `SByte`.

File: src/Datadog.Trace.ClrProfiler.Managed/InterceptMethodAttribute.cs [hunk 12]
```diff
@@ -59,6 +59,15 @@ public class InterceptMethodAttribute : Attribute
         /// </summary>
         public string TargetSignature { get; set; }
 
+        /// <summary>
+        /// Gets or sets the explicit type array for the target method to be intercepted.
+        /// This is a required field.
+        /// Must match the wrapper method in count:
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Managed/InterceptMethodAttribute.cs]: Is there additional documentation describing the order of the elements in this type array? i.e. return type followed by target method parameter types

File: test/Datadog.Trace.ClrProfiler.Managed.Tests/IntegrationSignatureTests.cs [hunk 13]
```diff
@@ -1,37 +1,74 @@
...
+using Datadog.Trace.ClrProfiler.Integrations;
 using Xunit;
 
 namespace Datadog.Trace.ClrProfiler.Managed.Tests
 {
     public class IntegrationSignatureTests
     {
-        [SuppressMessage("StyleCop.CSharp.SpacingRules", "SA1009:ClosingParenthesisMustBeSpacedCorrectly", Justification = "Reviewed.")]
-        public static IEnumerable<object[]> GetWrapperMethods()
+        private static readonly List<MethodInfo> StaticInstrumentations = new List<MethodInfo>()
```
Comment (zacharycmontoya) [inline test/Datadog.Trace.ClrProfiler.Managed.Tests/IntegrationSignatureTests.cs]: I feel like you could get the static methods dynamically so that, if we add new static wrapper methods, we don't have to remember to add them here.

Comment (zacharycmontoya) [reply test/Datadog.Trace.ClrProfiler.Managed.Tests/IntegrationSignatureTests.cs]: I think I spoke too soon without realizing you're trying to limit this to just the AspNetCoreMvc2Integration. It may still be beneficial to only have to list the class names and then we'd dynamically get all of the (publicly accessible) integration methods from that class.

Comment (colin-higgins) [reply test/Datadog.Trace.ClrProfiler.Managed.Tests/IntegrationSignatureTests.cs]: All of our methods are static, regardless of whether the target is. 
I think we would need to decorate whether the target is static in the attribute.

Comment (colin-higgins) [reply test/Datadog.Trace.ClrProfiler.Managed.Tests/IntegrationSignatureTests.cs]: At the very least, the arg counts need to match in the profiler for the instrumentation to happen, and if it's a static method, then this test will automatically fail until it's added to this list.

File: test/Datadog.Trace.ClrProfiler.Native.Tests/integration_loader_test.cpp [hunk 14]
```diff
@@ -204,3 +204,22 @@ TEST(IntegrationLoaderTest, LoadsFromEnvironment) {
...
+                "target": { "assembly": "Assembly.One", "type": "Type.One", "method": "Method.One", "signature_types": ["System.Void", null, "FakeClient.Pipeline'1<T>"] },
+                "wrapper": { "assembly": "Assembly.Two", "type": "Type.Two", "method": "Method.One", "signature": [0, 1, 1, 28] }
+            }]
+        }]
+    )TEXT");
+
+  auto integrations = LoadIntegrationsFromStream(str);
+  const auto target = integrations[0].method_replacements[0].target_method;
+  EXPECT_STREQ(L"System.Void", target.signature_types[0].c_str());
+  EXPECT_STREQ(L"System.Object", target.signature_types[1].c_str());
```
Comment (zacharycmontoya) [inline test/Datadog.Trace.ClrProfiler.Native.Tests/integration_loader_test.cpp]: How do we get System.Object from the input JSON?

Comment (colin-higgins) [reply test/Datadog.Trace.ClrProfiler.Native.Tests/integration_loader_test.cpp]: Refresh your page. ;)

File: test/Datadog.Trace.ClrProfiler.Managed.Tests/TypeNameTests.cs [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+        [Theory]
+        [MemberData(nameof(GetConstTypeAssociations))]
+        public void StringMatches(string constant, object type)
+        {
+            if (type is Type)
+            {
+                Assert.Equal(constant, ((Type)type).FullName);
+            }
+
+            if (type is string)
```
Comment (zacharycmontoya) [inline test/Datadog.Trace.ClrProfiler.Managed.Tests/TypeNameTests.cs]: When does this happen? I assume type would always be a System.Type object.

Comment (colin-higgins) [reply test/Datadog.Trace.ClrProfiler.Managed.Tests/TypeNameTests.cs]: It was happening when I was using UInt8 instead of Byte.
I may add the Ignore property in there though for full representation.
